File: src/main/java/com/teamwizardry/wizardry/api/capability/spell/SpellCapability.java
package com.teamwizardry.wizardry.api.capability.spell;

import com.teamwizardry.librarianlib.foundation.capability.BaseCapability;
import com.teamwizardry.librarianlib.prism.Save;
import com.teamwizardry.wizardry.common.spell.component.ShapeChain;

import net.minecraftforge.common.capabilities.Capability;
import net.minecraftforge.common.capabilities.CapabilityInject;

public class SpellCapability extends BaseCapability implements ISpellCapability
{
    @CapabilityInject(ISpellCapability.class)
    public static Capability<ISpellCapability> SPELL_CAPABILITY;
    
    @Save private ShapeChain spell;
    
    public SpellCapability(ShapeChain spell) { this.spell = spell; }
    
    @Override public ShapeChain getSpell() { return this.spell; }
    @Override public void setSpell(ShapeChain spell) { this.spell = spell; }
}

File: src/main/kotlin/com/teamwizardry/wizardry/capability/spell/SpellCapability.kt
package com.teamwizardry.wizardry.capability.spell

import com.teamwizardry.librarianlib.scribe.Save
import com.teamwizardry.wizardry.common.init.ModCapabilities
import com.teamwizardry.wizardry.common.spell.component.ShapeChain
import dev.onyxstudios.cca.api.v3.item.ItemComponent
import net.minecraft.item.ItemStack
import net.minecraft.nbt.NbtCompound

class SpellCapability(val stack: ItemStack, @Save override var spell: ShapeChain? = null): ISpellCapability, ItemComponent(stack, ModCapabilities.SPELL) {
    companion object {
        var ItemStack.spell: ShapeChain?
            get() = ModCapabilities.SPELL.maybeGet(this).orElse(null)?.spell
            set(value) {
                ModCapabilities.SPELL.maybeGet(this).orElse(null)?.let {it.spell = value}
            }
    }
}